src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ObjectiveCMarshal.NativeAot.cs

```diff
@@ -51,20 +55,122 @@ internal static void ThrowPendingExceptionObject()
             }
         }
 
+        [RuntimeExport("ObjectiveCMarshalTryGetTaggedMemory")]
+        static bool TryGetTaggedMemory(IntPtr pObj, IntPtr* tagged)
+        {
+            // We are paused in the GC, so this is safe.
+            object obj = Unsafe.AsRef<object>((void*)&pObj);
+
+            if (!s_objects.TryGetValue(obj, out ObjcTrackingInformation? info))
+            {
+                return false;
+            }
+
+            *tagged = info._memory;
+            return true;
+        }
+
+        [RuntimeExport("ObjectiveCMarshalGetIsTrackedReferenceCallback")]
+        static IntPtr GetIsTrackedReferenceCallback()
+        {
+            return s_IsTrackedReferenceCallback;
+        }
+
+        [RuntimeExport("ObjectiveCMarshalGetOnEnteredFinalizerQueueCallback")]
+        static IntPtr GetOnEnteredFinalizerQueueCallback()
+        {
+            return s_OnEnteredFinalizerQueueCallback;
+        }
+
         private static bool TryInitializeReferenceTracker(
             delegate* unmanaged<void> beginEndCallback,
             delegate* unmanaged<IntPtr, int> isReferencedCallback,
             delegate* unmanaged<IntPtr, void> trackedObjectEnteredFinalization)
         {
-            throw new NotImplementedException();
+            if (!RuntimeImports.RhRegisterObjectiveCMarshalBeginEndCallback((IntPtr)beginEndCallback))
+            {
+                return false;
+            }
+
+            s_IsTrackedReferenceCallback = (IntPtr)isReferencedCallback;
+            s_OnEnteredFinalizerQueueCallback = (IntPtr)trackedObjectEnteredFinalization;
+            s_initialized = true;
+
+            return true;
         }
 
         private static IntPtr CreateReferenceTrackingHandleInternal(
             object obj,
             out int memInSizeT,
             out IntPtr mem)
         {
-            throw new NotImplementedException();
+            if (!s_initialized)
+            {
+                throw new InvalidOperationException(SR.InvalidOperation_ObjectiveCMarshalNotInitialized);
+            }
+
+            if (!obj.GetEETypePtr().IsTrackedReferenceWithFinalizer)
+            {
+                throw new InvalidOperationException(SR.InvalidOperation_ObjectiveCTypeNoFinalizer);
+            }
+
+            var trackerInfo = s_objects.GetValue(obj, static o => new ObjcTrackingInformation());
+            trackerInfo.EnsureInitialized(obj);
+            trackerInfo.GetTaggedMemory(out memInSizeT, out mem);
+            return RuntimeImports.RhHandleAllocRefCounted(obj);
+        }
+
+        class ObjcTrackingInformation
+        {
+            const int TAGGED_MEMORY_SIZE_IN_POINTERS = 2;
```

Comment:
Add a note that this matches details in CoreCLR and should be kept in sync with that value.

Reply:
I added comments in both the CoreCLR and NativeAOT version of these constants referencing each other.